lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -610,35 +632,71 @@ diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl) {
+                      isa<ConstructorDecl>(decl) ? "initializer" :
+                      isa<FuncDecl>(decl) ? "method" :
+                      "member");
 
-      return tc.diagnose(parentDecl, diag::note_typo_candidate_implicit_member,
-                         decl->getBaseName().userFacingName(), kind);
-    }
+    return tc.diagnose(parentDecl, diag::note_typo_candidate_implicit_member,
+                       decl->getBaseName().userFacingName(), kind);
   }
 
   return tc.diagnose(decl, diag::note_typo_candidate,
                      decl->getBaseName().userFacingName());
 }
 
-void TypeChecker::noteTypoCorrection(DeclName writtenName, DeclNameLoc loc,
-                                     ValueDecl *decl) {
-  auto &&diagnostic = diagnoseTypoCorrection(*this, loc, decl);
+void TypoCorrectionResults::noteAllCandidates() const {
+  for (auto candidate : Candidates) {
+    auto &&diagnostic = noteTypoCorrection(TC, Loc, candidate);
 
-  DeclName declName = decl->getFullName();
+    TypoCorrection correction(WrittenName, Loc, candidate->getFullName());
+    correction.addFixits(diagnostic);
+  }
+}
 
-  if (writtenName.getBaseName() != declName.getBaseName())
-    diagnostic.fixItReplace(loc.getBaseNameLoc(),
-                            declName.getBaseName().userFacingName());
+void TypoCorrection::addFixits(InFlightDiagnostic &diagnostic) const {
+  if (WrittenName.getBaseName() != CorrectedName.getBaseName())
+    diagnostic.fixItReplace(Loc.getBaseNameLoc(),
+                            CorrectedName.getBaseName().userFacingName());
 
   // TODO: add fix-its for typo'ed argument labels.  This is trickier
   // because of the reordering rules.
 }
+
+Optional<TypoCorrection> TypoCorrectionResults::claimUniqueCorrection() {
+  // Look for a unique name.
+  DeclName uniqueCorrectedName;
+  for (auto candidate : Candidates) {
+    // If the candidate is implicit in a way that would require a more
+    // elaborate diagnostic to describe, it's not simple enough for this.
+    if (findExplicitParentForImplicitDecl(candidate))
+      return None;
+
+    auto candidateName = candidate->getFullName();
+
+    // If this is the first name, record it.
+    if (!uniqueCorrectedName)
+      uniqueCorrectedName = candidateName;
+
+    // If this is a different name from the last candidate, we don't have
+    // a unique correction.
+    else if (uniqueCorrectedName != candidateName)
+      return None;
+  }
+
+  // If we didn't find any candidates, we're done.
+  if (!uniqueCorrectedName)
+    return None;
+
+  // If the corrected name doesn't differ from the written name in its base
+  // name, it's not simple enough for this (for now).
+  if (WrittenName.getBaseName() == uniqueCorrectedName.getBaseName())
+    return None;
+
+  // Discard all the existing candidates so that the caller can just call
+  // noteAllCorrections().
+  Candidates.clear();
+
+  return TypoCorrection(WrittenName, Loc, uniqueCorrectedName.getBaseName());
```

Comment:
I'm a bit surprised that `TypoCorrection` doesn't include the candidate itself. The clients truly don't even bother to try to use that result to continue meaningfully?

Reply:
Oh, I guess they'll do the lookup again with the corrected name, which is likely safer.

Reply:
It's more that TypoCorrection is supposed to be basically "everything you would add to the diagnostic" more than it's meant to be a semantically self-describing thing, with the expectation that this will make it really easy to adopt typo corrections into diagnostics without repeating a lot of stuff.  For example, we actually immediately break down the DeclName of the candidate to just the base name on initialization because we currently never typo-correct anything about the argument labels.

You're right that having the candidate there would help with recovery.  On the other hand, promising to have a candidate there would prevent us from using this path in cases where there's a unique typo correction but it resolves to an overload set.